File: scripts/simpleKalman.py
#!/usr/bin/env python
"""
Sonar data processing using Kalman filter.

This script processes raw sonar data using a Kalman filter to estimate position 
and velocity. Results are saved to CSV files for further analysis.

Author: Sandy Herho (sandy.herho@email.ucr.edu)
Date: January 12, 2025
"""

import numpy as np
from scipy import io
import os
from pathlib import Path
import pandas as pd

class KalmanFilter:
    """Implementation of Kalman filter for position and velocity estimation."""
    
    def __init__(self):
        self.dt = 0.02  # Time step (seconds)
        
        # State transition matrix
        self.A = np.array([[1, self.dt],
                          [0, 1]])
        
        # Measurement matrix
        self.H = np.array([[1, 0]])
        
        # Process noise covariance
        self.Q = np.array([[1, 0],
                          [0, 3]])
        
        # Measurement noise covariance
        self.R = np.array([[10]])
        
        # Initial state
        self.x = np.array([[0],
                          [20]])
        
        # Initial state covariance
        self.P = 5 * np.eye(2)

    def update(self, measurement):
        """
        Update state estimate using Kalman filter algorithm.
        
        Args:
            measurement (np.ndarray): Current measurement value
            
        Returns:
            tuple: Position, velocity, and covariance matrix
        """
        # Prediction
        x_pred = self.A @ self.x
        P_pred = self.A @ self.P @ self.A.T + self.Q

        # Kalman gain
        S = self.H @ P_pred @ self.H.T + self.R
        K = P_pred @ self.H.T @ np.linalg.inv(S)

        # Update
        self.x = x_pred + K @ (measurement - self.H @ x_pred)
        self.P = P_pred - K @ self.H @ P_pred

        return self.x[0, 0], self.x[1, 0], self.P

class SonarData:
    """Handler for sonar measurement data."""
    
    def __init__(self, filename):
        """
        Load sonar data from MATLAB file.
        
        Args:
            filename (str): Path to the MATLAB data file
        """
        mat_data = io.loadmat(filename)
        self.measurements = mat_data['sonarAlt'].flatten()
        self.current_idx = 0
    
    def get_measurement(self):
        """Get next measurement from the dataset."""
        if self.current_idx >= len(self.measurements):
            return None
        measurement = self.measurements[self.current_idx]
        self.current_idx += 1
        return measurement

def ensure_dir(directory):
    """Create directory if it doesn't exist."""
    Path(directory).mkdir(parents=True, exist_ok=True)

def process_data():
    """Process sonar data and save results."""
    # Setup paths
    base_dir = Path(__file__).parent.parent
    raw_data_dir = base_dir / 'raw_data'
    processed_data_dir = base_dir / 'processed_data'
    
    # Ensure output directory exists
    ensure_dir(processed_data_dir)
    
    # Parameters
    n_samples = 1500
    dt = 0.02
    t = np.arange(0, n_samples * dt, dt)

    # Initialize arrays
    X_saved = np.zeros((n_samples, 2))  # Position and velocity
    Z_saved = np.zeros(n_samples)       # Measurements
    P_saved = np.zeros((n_samples, 2))  # State covariances

    # Process data
    sonar = SonarData(raw_data_dir / 'SonarAlt.mat')
    kf = KalmanFilter()

    for k in range(n_samples):
        z = sonar.get_measurement()
        if z is None:
            break
            
        pos, vel, P = kf.update(np.array([[z]]))

        X_saved[k, :] = [pos, vel]
        Z_saved[k] = z
        P_saved[k, :] = [P[0, 0], P[1, 1]]
    
    # Calculate velocities using finite differences
    vel_meas = np.diff(Z_saved) / dt
    vel_pos = np.diff(X_saved[:, 0]) / dt
    t_vel = t[:-1]
    
    # Save results
    # Main results
    df_main = pd.DataFrame({
        'time': t,
        'raw_position': Z_saved,
        'kalman_position': X_saved[:, 0],
        'kalman_velocity': X_saved[:, 1]
    })
    df_main.to_csv(processed_data_dir / 'kalman_estimates.csv', index=False)
    
    # Covariance data
    df_cov = pd.DataFrame({
        'time': t,
        'position_variance': P_saved[:, 0],
        'velocity_variance': P_saved[:, 1]
    })
    df_cov.to_csv(processed_data_dir / 'kalman_covariances.csv', index=False)
    
    # Velocity comparisons
    df_vel = pd.DataFrame({
        'time': t_vel,
        'measurement_velocity': vel_meas,
        'position_derivative': vel_pos,
        'kalman_velocity': X_saved[:-1, 1]
    })
    df_vel.to_csv(processed_data_dir / 'velocity_comparisons.csv', index=False)

if __name__ == "__main__":
    process_data()

File: scripts/simpleKalman.jl
#!/usr/bin/env julia
"""
Sonar data processing using Kalman filter.

This script processes raw sonar data using a Kalman filter to estimate position 
and velocity. Results are saved to CSV files for further analysis.

Author: Sandy Herho (sandy.herho@email.ucr.edu)
Date: January 12, 2025
License: WTFPL
"""

using MAT
using CSV
using DataFrames
using LinearAlgebra
using Dates

"""
Initialize Kalman filter parameters.

Returns:
    NamedTuple containing all necessary Kalman filter parameters
"""
function init_kalman_params()
    dt = 0.02  # Time step (seconds)
    
    # State transition matrix
    A = [1 dt; 0 1]
    
    # Measurement matrix
    H = [1 0]
    
    # Process noise covariance
    Q = [1 0; 0 3]
    
    # Measurement noise covariance
    R = [10]
    
    # Initial state
    x = [0; 20]
    
    # Initial state covariance
    P = 5 * Matrix(1.0I, 2, 2)
    
    return (;A, H, Q, R, x, P)
end

"""
Update Kalman filter state.

Args:
    params: Current Kalman filter parameters
    measurement: New measurement value

Returns:
    Tuple containing updated parameters and estimates (position, velocity, covariance)
"""
function kalman_update(params, measurement)
    # Unpack parameters
    (; A, H, Q, R, x, P) = params
    
    # Prediction
    x_pred = A * x
    P_pred = A * P * A' + Q
    
    # Kalman gain
    S = H * P_pred * H' .+ R
    K = P_pred * H' * inv(S)
    
    # Update
    x_new = x_pred + K * (measurement .- H * x_pred)
    P_new = P_pred - K * H * P_pred
    
    # Return updated parameters and estimates
    new_params = (; params..., x=x_new, P=P_new)
    pos = x_new[1]
    vel = x_new[2]
    
    return new_params, pos, vel, P_new
end

"""
Ensure directory exists, create if necessary.
"""
function ensure_dir(dir_path)
    isdir(dir_path) || mkpath(dir_path)
end

"""
Main processing function.
"""
function process_data()
    # Setup paths
    base_dir = dirname(dirname(abspath(@__FILE__)))
    raw_data_dir = joinpath(base_dir, "raw_data")
    processed_data_dir = joinpath(base_dir, "processed_data")
    
    # Ensure output directory exists
    ensure_dir(processed_data_dir)
    
    # Parameters
    n_samples = 1500
    dt = 0.02
    t = range(0, step=dt, length=n_samples)
    
    # Initialize arrays
    X_saved = zeros(n_samples, 2)  # Position and velocity
    Z_saved = zeros(n_samples)     # Measurements
    P_saved = zeros(n_samples, 2)  # State covariances
    
    # Load sonar data
    file = matread(joinpath(raw_data_dir, "SonarAlt.mat"))
    measurements = vec(file["sonarAlt"])
    
    # Process data
    params = init_kalman_params()
    
    for k in 1:n_samples
        # Get measurement
        z = measurements[k]
        
        # Update Kalman filter
        params, pos, vel, P = kalman_update(params, z)
        
        # Save results
        X_saved[k, :] = [pos, vel]
        Z_saved[k] = z
        P_saved[k, :] = [P[1,1], P[2,2]]
    end
    
    # Calculate velocities using finite differences
    vel_meas = diff(Z_saved) ./ dt
    vel_pos = diff(X_saved[:, 1]) ./ dt
    t_vel = t[1:end-1]
    
    # Save results to CSV
    # Main results
    df_main = DataFrame(
        time = t,
        raw_position = Z_saved,
        kalman_position = X_saved[:, 1],
        kalman_velocity = X_saved[:, 2]
    )
    CSV.write(joinpath(processed_data_dir, "kalman_estimates_jl.csv"), df_main)
    
    # Covariance data
    df_cov = DataFrame(
        time = t,
        position_variance = P_saved[:, 1],
        velocity_variance = P_saved[:, 2]
    )
    CSV.write(joinpath(processed_data_dir, "kalman_covariances_jl.csv"), df_cov)
    
    # Velocity comparisons
    df_vel = DataFrame(
        time = t_vel,
        measurement_velocity = vel_meas,
        position_derivative = vel_pos,
        kalman_velocity = X_saved[1:end-1, 2]
    )
    CSV.write(joinpath(processed_data_dir, "velocity_comparisons_jl.csv"), df_vel)
end

# Run the processing if this is the main script
if abspath(PROGRAM_FILE) == @__FILE__
    process_data()
end
